pinot-core/src/main/java/com/linkedin/pinot/core/operator/transform/DateTimeConversionEvaluator.java

```diff
@@ -0,0 +1,213 @@
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *         http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package com.linkedin.pinot.core.operator.transform;
+
+import java.util.concurrent.TimeUnit;
+
+import org.apache.commons.lang3.EnumUtils;
+import org.joda.time.DateTime;
+import org.joda.time.DateTimeZone;
+import org.joda.time.Days;
+import org.joda.time.Hours;
+import org.joda.time.Interval;
+import org.joda.time.Minutes;
+import org.joda.time.Months;
+import org.joda.time.MutableDateTime;
+import org.joda.time.Seconds;
+import org.joda.time.Weeks;
+import org.joda.time.Years;
+import org.joda.time.format.DateTimeFormat;
+import org.joda.time.format.DateTimeFormatter;
+
+import com.google.common.base.Preconditions;
+import com.linkedin.pinot.common.data.DateTimeFieldSpec.TimeFormat;
+import com.linkedin.pinot.common.utils.time.DateTimeFieldSpecUtils;
+
+/**
+ * Converts the time column value based on the inputFormat, outputFormat and outputGranularity
+ */
+public class DateTimeConversionEvaluator {
+
+  private static final DateTime EPOCH_START_DATE = new DateTime(0L, DateTimeZone.UTC);
+
+  public enum DateTimeTransformUnit {
+    YEARS,
+    MONTHS,
+    WEEKS,
+    DAYS,
+    HOURS,
+    MINUTES,
+    SECONDS,
+    MILLISECONDS;
+  }
+
+  private String inputFormat;
+  private String outputFormat;
+  private String outputGranularity;
+  private DateTimeFormatter inputDateTimeFormatter = null;
+  private DateTimeFormatter outputDateTimeFormatter = null;
+
+  public DateTimeConversionEvaluator(String inputFormat, String outputFormat, String outputGranularity) {
+
+    Preconditions.checkArgument(DateTimeFieldSpecUtils.isValidFormat(inputFormat));
+    Preconditions.checkArgument(isValidFormat(outputFormat));
+    Preconditions.checkArgument(DateTimeFieldSpecUtils.isValidGranularity(outputGranularity));
+    this.inputFormat = inputFormat;
+    this.outputFormat = outputFormat;
+    this.outputGranularity = outputGranularity;
+    String inputSDFFormat = DateTimeFieldSpecUtils.getSDFPatternFromFormat(inputFormat);
+    if (inputSDFFormat != null) {
+      inputDateTimeFormatter = DateTimeFormat.forPattern(inputSDFFormat).withZoneUTC();
+    }
+    String outputSDFFormat = DateTimeFieldSpecUtils.getSDFPatternFromFormat(outputFormat);
+    if (outputSDFFormat != null) {
+      outputDateTimeFormatter = DateTimeFormat.forPattern(outputSDFFormat).withZoneUTC();
+    }
+  }
+
+  /**
+   * Transforms the given input value into the new format and bucket
+   * @param inputValue
+   * @return
+   */
+  public long transformDateTime(Object inputValue) {
```

Comment:
these can be optimized further. We should set up everything in the constructor so that the processing for each value is low.